### PR TITLE
Add a second network interface to VMs

### PR DESCRIPTION
**Pull request by scabrero**
Changes:

* Rename configuration variables to use `BRx_` and `VMx_` as prefix
* Allow to skip the bridge device provision with `BRx_DEV_SKIP_PROVISION`, useful to reuse the bridge devices created by libvirt and connect the rapido VMs to them
* Allow to create a second network adapter in each VM, connected to the second bridge device `BR2_DEV`.
* Setup the network in `vm_autorun.env` instead of using the `ip=` kernel parameter. It will allow to use IPv6 addresses in the future.

### CODE REVIEW

File: tools/br_setup.sh [hunk 1]
```diff
@@ -15,63 +15,145 @@
...
+		echo -n " with address $BR1_ADDR"
+	fi
+
+	if [ -n "$BR1_IF" ]; then
+		ip link set $BR1_IF master $BR1_DEV || exit 1
+		unwind="ip link set $BR1_IF nomaster; ${unwind}"
+		echo -n ", connected to $BR1_IF"
+	fi
+	echo
 fi
```
Comment (ddiss) [inline tools/br_setup.sh:41]: Instead of adding the `BR1_DEV_SKIP_PROVISON` param, could we just ignore any preexisting bridge interface here?

**Review by ddiss — CHANGES_REQUESTED** [review]
Thanks a lot for your changes here, Samuel.

Aside from the minor inline comments I have two concerns regarding the approach:
- although very helpful for some use-cases, I'm worried about the configuration complexity added for new users here.
  + perhaps we could reuse some other existing config syntax (sysconfig ifcfg or dnsmasq) so that users don't need to learn something new, and we don't need to continuously to add new rapido.conf options
  + I raised this approach a while back via https://github.com/rapido-linux/rapido/issues/89
- this will break users' with existing rapido.conf network configs, requiring them to rewrite them.
  + I don't want to support old and new syntax, but maybe some sort of upgrade helper would be worth while?
- moving from kernel dhcp client to `dhclient` seem like a good idea (finally we can support standard distro kernels), but it does appear to add a little bit of bloat to images

**Comment by ddiss** [comment]
> Thanks a lot for your changes here, Samuel.
> 
> Aside from the minor inline comments I have some concerns regarding the approach:
> 
>     * although very helpful for some use-cases, I'm worried about the configuration complexity added for new users here.
>       
>       * perhaps we could reuse some other existing config syntax (sysconfig ifcfg or dnsmasq) so that users don't need to learn something new, and we don't need to continuously to add new rapido.conf options
>       * I raised this approach a while back via #89
> 
>     * this will break users' with existing rapido.conf network configs, requiring them to rewrite them.
>       
>       * I don't want to support old and new syntax, but maybe some sort of upgrade helper would be worth while?
> 
>     * moving from kernel dhcp client to `dhclient` seem like a good idea (finally we can support standard distro kernels), but it does appear to add a little bit of bloat to images

@scabrero any thoughts on how you think we should proceed here?

File: tools/br_setup.sh (hunk 1 above)
Comment (scabrero) [reply tools/br_setup.sh:41]: This is a good idea. I will change it.

**Comment by scabrero** [comment]
> > Thanks a lot for your changes here, Samuel.
> > Aside from the minor inline comments I have some concerns regarding the approach:
> > ```
> > * although very helpful for some use-cases, I'm worried about the configuration complexity added for new users here.
> >   
> >   * perhaps we could reuse some other existing config syntax (sysconfig ifcfg or dnsmasq) so that users don't need to learn something new, and we don't need to continuously to add new rapido.conf options
> >   * I raised this approach a while back via #89
> > 
> > * this will break users' with existing rapido.conf network configs, requiring them to rewrite them.
> >   
> >   * I don't want to support old and new syntax, but maybe some sort of upgrade helper would be worth while?
> > 
> > * moving from kernel dhcp client to `dhclient` seem like a good idea (finally we can support standard distro kernels), but it does appear to add a little bit of bloat to images
> > ```
> 
> @scabrero any thoughts on how you think we should proceed here?

I had a look to #89 and I am wondering, what do you think if we add systemd and systemd-networkd to all images and provide the systemd-networkd files for each vm in `rapido-net-conf/vmX/`, injecting them into the initramfs image using DRACUT_RAPIDO_INCLUDES?

We could extend it and install systemd-resolved too to have DNS resolution in the VMs.

**Comment by ddiss** [comment]
> I had a look to #89 and I am wondering, what do you think if we add systemd and systemd-networkd to all images and provide the systemd-networkd files for each vm in rapido-net-conf/vmX/, injecting them into the initramfs image using DRACUT_RAPIDO_INCLUDES?

I'd like to keep dependencies/images minimal, so using systemd-networkd only would be my preference if possible (it's under 500k and appears to be quite powerful). I had a play with it myself, but couldn't get it to bring up a dhcp client address by itself. In the end all I really want is an clean network config syntax which could be used for configuring an arbitrary number of NICs per-VM, with static or dhcp based v4 or v6 addresses, etc., preferably outside of rapido.conf .

**Comment by ddiss** [comment]
> I'd like to keep dependencies/images minimal, so using systemd-networkd only would be my preference if possible (it's under 500k and appears to be quite powerful). I had a play with it myself, but couldn't get it to bring up a dhcp client address by itself. In the end all I really want is an clean network config syntax which could be used for configuring an arbitrary number of NICs per-VM, with static or dhcp based v4 or v6 addresses, etc., preferably outside of rapido.conf .

Just an update here - I plan on retesting with an updated systemd-networkd once I upgrade to openSUSE 15.3. @scabrero given that you had systemd-networkd working standalone on Tumbleweed, I'm hoping it's just the newer version that's needed before we can merge here.

**Comment by ddiss** [comment]
> Just an update here - I plan on retesting with an updated systemd-networkd once I upgrade to openSUSE 15.3. @scabrero given that you had systemd-networkd working standalone on Tumbleweed, I'm hoping it's just the newer version that's needed before we can merge here.

The systemd-networkd approach tracked via https://github.com/rapido-linux/rapido/pull/147 seems to be progressing quite well, and should offer a cleaner and more flexible configuration format supporting multihoming. @scabrero I'm going to close this, but feel free to reopen if you feel strongly about keeping these changes.

**Pull request closed without merging**